e2e/utils/visual-regression.js

```diff
@@ -174,3 +202,95 @@ module.exports.getRefPicName = () => {
+  }
+  return basePathRef + `${slugify(parent.toLowerCase())}/${slugify(testName.toLowerCase())}/${lastWordOfTestName.toLowerCase()}_reference_pic.png`;
+}
+
+function handleTakenScreenshot(data, misMatchTolerance, selector) {
+  const isTestPassed = (data.misMatchPercentage < misMatchTolerance) || false;
+  return new Promise((resolve, reject) => {
+    if (!isTestPassed) {
+      mkdirp.sync(basePathDiff + `${slugify(browser.currentDescribe.toLowerCase())}/${slugify(browser.currentTestName.toLowerCase())}`);
+
+      data
+        .getDiffImage()
+        .pack()
+        .on('error', (err) => reject(err))
+        .on('end', () => {
+          resolve(isTestPassed);
+        })
+        .pipe(
+          fs.createWriteStream(basePathDiff + createTestName())
+        );
+
+      testDebug('failing testName: ' + browser.currentTestName
+      + '\nbrowser: ' + browser.desiredCapabilities.browserName
+      + '\nplatform: ' + browser.desiredCapabilities.platform
+      + '\nmisMatchTolerance: ' + misMatchTolerance
+      + '\nmisMatchPercentage: ' + data.misMatchPercentage
+      + '\nproblematic elementSelector: ' + selector);
+      return;
+    }
+    resolve(isTestPassed);
+  });
+}
+function writeDataIfNeeded(img, misMatchTolerance, selector) {
+  return new Promise((resolve, reject) => {
+    img.onComplete(function (data) {
+      handleTakenScreenshot(data, misMatchTolerance, selector)
+        .then(res => resolve(res))
+        .catch(err => reject(err));
+    });
+  });
+}
+module.exports.takeScreenShotOfElementAndCompareWithRef = (selector, options) => {
+  return module.exports.takeScreenShotOfElementWithCrop(selector, options).value;
+};
+
+module.exports.takeScreenShotOfElementWithCrop = (selector, options) => {
+  let misMatchTolerance = options.defaultTolerance;
+
+  setBasePath2(path.dirname(browser.currentTest));
+  const testPathAndName = basePathScreen + createTestName();
+
+  if (options.windowsTolerance
+    && browser.desiredCapabilities.platform
+    && browser.desiredCapabilities.platform.includes('Windows')) {
+    misMatchTolerance = options.windowsTolerance;
+  }
+
+  if (options.firefoxTolerance
+    && browser.desiredCapabilities.browserName
+    && browser.desiredCapabilities.browserName.includes('firefox')) {
+    misMatchTolerance = options.firefoxTolerance;
+  }
+
+  if (options.ffAndWindowsTolerance
+  && browser.desiredCapabilities.browserName
+  && browser.desiredCapabilities.browserName.includes('firefox')) {
+    misMatchTolerance = options.ffAndWindowsTolerance;
+  }
+
+  browser.saveElementScreenshot(selector, testPathAndName);
+  const img = resemble(getRefPicName()).compareTo(testPathAndName);
+  if (options.ignoreComparison === true) {
+    img.ignoreColors();
+  }
+
+  const promise = writeDataIfNeeded(img, misMatchTolerance, selector).then(result => {
+    return {
+      value: result
+    };
+  });
+  return browser.waitUntil(() => {
+    return promise;
+  }, 1e4);
```

Comment:
shall we keep this 1e4? Is it the right amount of time to wait?

Reply:
I think we can increase it to 1e5, it should be enough.

Reply:
let's leave this, 10 sec should more than enough.